dashboards/cls_dashboard.py

```diff
@@ -0,0 +1,159 @@
+import torch
+import numpy as np
+import seaborn as sns
+import streamlit as st
+import matplotlib.pyplot as plt
+from datasets import load_dataset
+from sklearn.decomposition import PCA
+from sklearn.tree import DecisionTreeClassifier
+from torch.utils.data import Subset, DataLoader
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import confusion_matrix
+from transformers import RobertaConfig, RobertaForMaskedLM
+
+from data.dataset import MidiDataset
+from data.quantizer import MidiQuantizer
+from data.tokenizer import QuantizedMidiEncoder
+
+
+def preprocess_dataset(
+    dataset_name: str,
+    quantizer: MidiQuantizer,
+    tokenizer: QuantizedMidiEncoder,
+    queries: list[str],
+    batch_size: int,
+    num_workers: int,
+):
+    dataset = load_dataset(dataset_name, split="validation")
+
+    ds = MidiDataset(dataset, quantizer, tokenizer, masking_probability=0.0)
+
+    idx = []
+
+    for query in queries:
+        idx_query = [i for i, name in enumerate(ds.dataset["source"]) if str.lower(query) in str.lower(name)]
+        idx += idx_query
+
+    ds = Subset(ds, indices=idx)
+
+    # dataloaders
+    dataloader = DataLoader(ds, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+
+    return dataloader
+
+
+def plot_embeddings(X: np.ndarray, labels: list[str]):
+    pca = PCA(n_components=2)
+    X_reduced = pca.fit_transform(X)
+
+    fig = plt.figure(figsize=(10, 10))
+    sns.scatterplot(x=X_reduced[:, 0], y=X_reduced[:, 1], hue=labels)
+    st.pyplot(fig)
+
+def plot_confusion_matrix(cm: np.ndarray):
+    fig = plt.figure(figsize=(10, 10))
+    sns.heatmap(cm, annot=True)
+    st.pyplot(fig)
+
+
+def evaluate_classification(
+    model: RobertaForMaskedLM,
+    dataset_name: str,
+    quantizer: MidiQuantizer,
+    tokenizer: QuantizedMidiEncoder,
+    queries: list[str],
+    device: torch.device,
+):
+    dataloader = preprocess_dataset(
+        dataset_name,
+        quantizer=quantizer,
+        tokenizer=tokenizer,
+        queries=queries,
+        batch_size=1024,
+        num_workers=8,
+    )
+    query_to_id = {k: i for i, k in enumerate(queries)}
+
+    outputs = []
+    labels = []
+
+    # velocity time encoding
+    for batch in dataloader:
+        with torch.no_grad():
+            source = batch["source"]
+            input_token_ids = batch["input_token_ids"].to(device)
+
+            # shape [batch_size, seq_len, vocab_size]
+            out = model(input_ids=input_token_ids).logits
+            cls_embedding = out[:, 0, :]
+            outputs.append(cls_embedding)
+
+            for s in source:
+                for query in queries:
+                    if str.lower(query) in str.lower(s):
+                        labels.append(query)
```

Comment:
It doesn't make sense to process everything and only later look for records matching the queries. Can you refactor it into steps more like `query -> dataset -> model -> PCA`?